src/naive_bayes/bernoulli.rs

```diff
@@ -0,0 +1,278 @@
+            class_priors,
+            feature_prob,
+        })
+    }
+}
+
+/// BernoulliNB implements the categorical naive Bayes algorithm for categorically distributed data.
+#[derive(Serialize, Deserialize, Debug, PartialEq)]
+pub struct BernoulliNB<T: RealNumber, M: Matrix<T>> {
+    inner: BaseNaiveBayes<T, M, BernoulliNBDistribution<T>>,
+    binarize: Option<T>,
+}
+
+impl<T: RealNumber, M: Matrix<T>> BernoulliNB<T, M> {
+    /// Fits BernoulliNB with given data
+    /// * `x` - training data of size NxM where N is the number of samples and M is the number of
+    /// features.
+    /// * `y` - vector with target values (classes) of length N.
+    /// * `parameters` - additional parameters like class priors, alpha for smoothing and
+    /// binarizing threshold.
+    pub fn fit(
+        x: &M,
+        y: &M::RowVector,
+        parameters: BernoulliNBParameters<T>,
+    ) -> Result<Self, Failed> {
+        let distribution = if let Some(threshold) = parameters.binarize {
+            BernoulliNBDistribution::fit(
+                &binarize(x, threshold),
+                y,
+                parameters.alpha,
+                parameters.priors,
+            )?
+        } else {
+            BernoulliNBDistribution::fit(x, y, parameters.alpha, parameters.priors)?
+        };
+
+        let inner = BaseNaiveBayes::fit(distribution)?;
+        Ok(Self {
+            inner,
+            binarize: parameters.binarize,
+        })
+    }
+
+    /// Estimates the class labels for the provided data.
+    /// * `x` - data of shape NxM where N is number of data points to estimate and M is number of features.
+    /// Returns a vector of size N with class estimates.
+    pub fn predict(&self, x: &M) -> Result<M::RowVector, Failed> {
+        if let Some(threshold) = self.binarize {
+            self.inner.predict(&binarize(x, threshold))
+        } else {
+            self.inner.predict(x)
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::linalg::naive::dense_matrix::DenseMatrix;
+
+    #[test]
+    fn run_bernoulli_naive_bayes() {
+        // Tests that BernoulliNB when alpha=1.0 gives the same values as
+        // those given for the toy example in Manning, Raghavan, and
+        // Schuetze's "Introduction to Information Retrieval" book:
+        // https://nlp.stanford.edu/IR-book/html/htmledition/the-bernoulli-model-1.html
+
+        // Training data points are:
+        // Chinese Beijing Chinese (class: China)
+        // Chinese Chinese Shanghai (class: China)
+        // Chinese Macao (class: China)
+        // Tokyo Japan Chinese (class: Japan)
+        let x = DenseMatrix::<f64>::from_2d_array(&[
+            &[1., 1., 0., 0., 0., 0.],
+            &[0., 1., 0., 0., 1., 0.],
+            &[0., 1., 0., 1., 0., 0.],
+            &[0., 1., 1., 0., 0., 1.],
+        ]);
+        let y = vec![0., 0., 0., 1.];
+        let bnb = BernoulliNB::fit(&x, &y, Default::default()).unwrap();
+
```

Comment:
Can you also add this test please? It doesn't change anything but it's good to have a direct comparison with Scikit. 

```
#[test]
    fn bernoulli_nb_scikit_parity() {
        let x = DenseMatrix::<f64>::from_2d_array(&[
            &[2., 4., 0., 0., 2., 1., 2., 4., 2., 0.],
            &[3., 4., 0., 2., 1., 0., 1., 4., 0., 3.],
            &[1., 4., 2., 4., 1., 0., 1., 2., 3., 2.],
            &[0., 3., 3., 4., 1., 0., 3., 1., 1., 1.],
            &[0., 2., 1., 4., 3., 4., 1., 2., 3., 1.],
            &[3., 2., 4., 1., 3., 0., 2., 4., 0., 2.],
            &[3., 1., 3., 0., 2., 0., 4., 4., 3., 4.],
            &[2., 2., 2., 0., 1., 1., 2., 1., 0., 1.],
            &[3., 3., 2., 2., 0., 2., 3., 2., 2., 3.],
            &[4., 3., 4., 4., 4., 2., 2., 0., 1., 4.],
            &[3., 4., 2., 2., 1., 4., 4., 4., 1., 3.],
            &[3., 0., 1., 4., 4., 0., 0., 3., 2., 4.],
            &[2., 0., 3., 3., 1., 2., 0., 2., 4., 1.],
            &[2., 4., 0., 4., 2., 4., 1., 3., 1., 4.],
            &[0., 2., 2., 3., 4., 0., 4., 4., 4., 4.]]);
        let y = vec![2., 2., 0., 0., 0., 2., 1., 1., 0., 1., 0., 0., 2., 0., 2.];
        let bnb = BernoulliNB::fit(&x, &y, Default::default()).unwrap();        

        let y_hat = bnb.predict(&x).unwrap();

        assert!(bnb.inner.distribution.class_priors.approximate_eq(&vec!(0.46, 0.2, 0.33), 1e-2));
        assert!(bnb.inner.distribution.feature_prob[1].approximate_eq(&vec!(0.8, 0.8, 0.8, 0.4, 0.8, 0.6, 0.8, 0.6, 0.6, 0.8), 1e-1));
        assert!(y_hat.approximate_eq(&vec!(2.0, 2.0, 0.0, 0.0, 0.0, 2.0, 1.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0), 1e-5));
    }
```